lib/Dialect/FIRRTL/Transforms/LowerAnnotations.cpp

```diff
@@ -690,29 +691,45 @@ LogicalResult LowerAnnotationsPass::solveWiringProblems(ApplyState &state) {
     moduleModifications[sinkModule].connectionMap[index] = sink;
 
     // Record ports that should be added to each module along the LCA path.
-    // Wire using RefType based on the source-- the RefSend will be of this
-    // type, and we cannot connect RefType's of non-identical types. The final
-    // connect of the resolved ref to the sink will handle any differences.
-    RefType tpe = TypeSwitch<Type, RefType>(problem.source.getType())
-                      .Case<FIRRTLBaseType>([](FIRRTLBaseType base) {
-                        return RefType::get(base);
-                      })
-                      .Case<RefType>([](RefType ref) { return ref; });
-    for (auto sourceInst : sources) {
+    RefType refType = TypeSwitch<Type, RefType>(problem.source.getType())
+                          .Case<FIRRTLBaseType>([](FIRRTLBaseType base) {
+                            return RefType::get(base);
+                          })
+                          .Case<RefType>([](RefType ref) { return ref; });
+    StringRef name;
+    for (auto sourceInst : llvm::reverse(sources)) {
       auto mod = cast<FModuleOp>(instanceGraph.getReferencedModule(sourceInst));
+      if (name.empty()) {
+        if (problem.newNameHint.empty())
+          name = state.getNamespace(mod).newName(
+              getFieldName(getFieldRefFromValue(source), /*nameSafe=*/true)
+                  .first +
+              "__bore");
+        else
+          name = state.getNamespace(mod).newName(problem.newNameHint);
+      }
       moduleModifications[mod].portsToAdd.push_back(
-          {index,
-           {StringAttr::get(
-                context, state.getNamespace(mod).newName(problem.newNameHint)),
-            tpe, Direction::Out}});
+          {index, {StringAttr::get(context, name), refType, Direction::Out}});
+      name = state.getNamespace(mod).newName(sourceInst.instanceName() + "_" +
+                                             name);
     }
-    for (auto sinkInst : sinks) {
+    name = "";
+    for (auto sinkInst : llvm::reverse(sinks)) {
       auto mod = cast<FModuleOp>(instanceGraph.getReferencedModule(sinkInst));
+      if (name.empty()) {
+        if (problem.newNameHint.empty())
+          name = state.getNamespace(mod).newName(
+              getFieldName(getFieldRefFromValue(sink), /*nameSafe=*/true)
+                  .first +
+              "__bore");
+        else
+          name = state.getNamespace(mod).newName(problem.newNameHint);
+      }
       moduleModifications[mod].portsToAdd.push_back(
           {index,
-           {StringAttr::get(
-                context, state.getNamespace(mod).newName(problem.newNameHint)),
-            tpe, Direction::In}});
+           {StringAttr::get(context, name), refType.getType(), Direction::In}});
```

Comment:
Wow `connect` is pretty magical that this works, awesome!

Don't think it's used this way today, but now if the sink is a reftype and we need to bore down to it, we'll 'send' from the bored 'real input' port added to that module directly to the sink (when doing the final connect from that port to the reftype sink)?

Not sure there's anything better to be done, other than detecting/rejecting such a situation but probably not what would be expected.

Related: if we don't might be good to canonicalize trivial send/resolve chains (if not resolve/send)...

Just something to keep in mind as this gets more users....